src/backend/columnar/columnar_tableam.c

```diff
@@ -585,18 +547,37 @@ columnar_index_fetch_tuple(struct IndexFetchTableData *sscan,
 	}
 	else if (StripeWriteInProgress(stripeMetadata))
 	{
-		/* similar to aborted writes .. */
-		Assert(snapshot->snapshot_type == SNAPSHOT_DIRTY);
+		if (stripeMetadata->insertedByCurrentXact)
+		{
+			/*
+			 * Stripe is visible to me, so it must be written by me. Since
+			 * caller might want to use tupleslot datums for some reason, do
+			 * another look-up, but this time by first flushing our writes.
+			 */
+			ColumnarReadFlushPendingWrites(scan->cs_readState);
```

Comment:
I know we talked about this this morning (yesterday for you), but why don't we need to read tuples from another transaction? It seems that there are a couple assumptions here:

1. If it's in-progress and by another transaction, we must be doing a dirty read
2. and we don't need to read the actual tuple because it's a btree unique check

Those may be true, but let's be more clear about the assumptions we are making and use some XXX to try to make it more solid later. It feels pretty hacky right now.

Reply:
done now